Applications/Spire/Source/Ui/FilteredDropDownMenu.cpp

```diff
@@ -0,0 +1,214 @@
+#include "Spire/Ui/FilteredDropDownMenu.hpp"
+#include <QKeyEvent>
+#include <QPainter>
+#include "Spire/Spire/Dimensions.hpp"
+
+using namespace Spire;
+
+namespace {
+  auto PADDING() {
+    static auto padding = scale_width(8);
+    return padding;
+  }
+}
+
+FilteredDropDownMenu::FilteredDropDownMenu(std::vector<QVariant> items,
+    QWidget* parent)
+    : TextInputWidget(parent),
+      m_was_click_focused(false) {
+  setAttribute(Qt::WA_Hover);
+  setFocusPolicy(Qt::StrongFocus);
+  connect(this, &QLineEdit::editingFinished, this,
+    &FilteredDropDownMenu::on_editing_finished);
+  connect(this, &QLineEdit::textEdited, this,
+    &FilteredDropDownMenu::on_text_edited);
+  if(!items.empty()) {
+    m_current_item = items.front();
+    setText(m_item_delegate.displayText(m_current_item));
+  }
+  m_menu_list = new DropDownList({}, false, this);
+  m_item_activated_connection = m_menu_list->connect_activated_signal(
+    [=] (const auto& value) { on_item_activated(value); });
+  m_list_selection_connection = m_menu_list->connect_value_selected_signal(
+    [=] (const auto& value) { on_item_selected(value, true); });
+  set_items(std::move(items));
+  m_menu_list->installEventFilter(this);
+}
+
+bool FilteredDropDownMenu::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == m_menu_list) {
+    if(event->type() == QEvent::Show) {
+      m_menu_list->setFixedWidth(width());
+      update();
+    }
+  }
+  return TextInputWidget::eventFilter(watched, event);
+}
+
+void FilteredDropDownMenu::focusInEvent(QFocusEvent* event) {
+  TextInputWidget::focusInEvent(event);
+  if(event->reason() == Qt::FocusReason::MouseFocusReason) {
+    m_was_click_focused = true;
+  }
+}
+
+void FilteredDropDownMenu::focusOutEvent(QFocusEvent* event) {
+  if(!m_menu_list->isActiveWindow()) {
+    TextInputWidget::focusOutEvent(event);
+  }
+}
+
+void FilteredDropDownMenu::keyPressEvent(QKeyEvent* event) {
+  switch(event->key()) {
+    case Qt::Key_Delete:
+      setText("");
+      return;
+    case Qt::Key_Escape:
+      setText(m_item_delegate.displayText(m_current_item));
+    case Qt::Key_Enter:
+    case Qt::Key_Return:
+      if(m_last_activated_item.isValid()) {
+        on_item_selected(m_last_activated_item, true);
+        return;
+      }
+      if(!text().isEmpty() && m_menu_list->isVisible()) {
+        auto item = m_menu_list->get_value(0);
+        if(item.isValid()) {
+          on_item_selected(item, true);
+        }
+      }
+      return;
+  }
+  TextInputWidget::keyPressEvent(event);
+}
+
+void FilteredDropDownMenu::mousePressEvent(QMouseEvent* event) {
+  TextInputWidget::mousePressEvent(event);
+  if(m_was_click_focused) {
+    selectAll();
+    m_was_click_focused = false;
+  }
+}
+
+void FilteredDropDownMenu::paintEvent(QPaintEvent* event) {
+  TextInputWidget::paintEvent(event);
+  if(!text().isEmpty()) {
+    auto item = m_menu_list->get_value(0);
```

Comment:
Can you use ```auto&```? What is the return type of ```get_value```?

Reply:
It isn't a problem to use auto& here because get_value(...) returns a const reference that will be valid for as long as item is being used.